<a href="https://colab.research.google.com/github/alexDx12/gb_statistics/blob/lesson_7/lesson_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Теория вероятностей и математическая статистика

## Урок 7

## Линейная регрессия. Двухвыборочный t-тест. A/B-тестирование

### Домашнее задание

### Задача 1

__Постановка задачи__

Дана матрица объект-признак
```
X = [[ 1.22401313,  2.30868478,  3.03636353,  2.69287214],
     [-0.18757272,  1.30337355,  5.12093014,  3.46363202],
     [-0.81094525,  1.82463398,  5.79686488,  1.86159445],
     [ 0.75129018,  2.67392052,  3.65529809,  1.66746094],
     [ 0.00972362,  1.97367255,  2.50594319,  1.69755173],
     [-0.62972637,  0.77750764,  2.84124027,  4.54410559],
     [ 2.29536229,  1.81206697,  1.95026215,  1.51874636],
     [ 0.0920418 ,  2.26971361,  7.47708735,  2.61081203],
     [ 2.39252799,  3.17563985,  3.61420599,  5.10773362],
     [ 0.54983815,  2.87988651,  1.65752765,  1.59635987]]
```
и значения целевой переменной
```
y = [ 9.26193358,  9.700363  ,  8.67214805,  8.74796974,  6.18689108,
      7.53312713,  7.57643777, 12.44965478, 14.29010746,  6.68361218]
```

Подберите два признака (из четырёх) так, чтобы уровень линейной зависимости целевой переменной от значений этих признаков был максимальным. Другими словами, модель линейной регрессии на этих признаках должна давать наилучший результат. В качестве ответа запишите значение коэффициента детерминации для модели на этих признаках.

__Решение__

Импорт библиотек и скриптов:

In [ ]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from scipy import stats

Матрица объект-признак (X) и значения целевой переменной (y) исходя из постановки задачи:

In [ ]:
X = [[ 1.22401313,  2.30868478,  3.03636353,  2.69287214],
     [-0.18757272,  1.30337355,  5.12093014,  3.46363202],
     [-0.81094525,  1.82463398,  5.79686488,  1.86159445],
     [ 0.75129018,  2.67392052,  3.65529809,  1.66746094],
     [ 0.00972362,  1.97367255,  2.50594319,  1.69755173],
     [-0.62972637,  0.77750764,  2.84124027,  4.54410559],
     [ 2.29536229,  1.81206697,  1.95026215,  1.51874636],
     [ 0.0920418 ,  2.26971361,  7.47708735,  2.61081203],
     [ 2.39252799,  3.17563985,  3.61420599,  5.10773362],
     [ 0.54983815,  2.87988651,  1.65752765,  1.59635987]]
y = [ 9.26193358,  9.700363  ,  8.67214805,  8.74796974,  6.18689108,
      7.53312713,  7.57643777, 12.44965478, 14.29010746,  6.68361218]

Транспонирование матрицы объект-признак X в матрицу XT с целью упрощения извлечения признаков:

In [ ]:
XT = np.transpose(X)
XT

array([[ 1.22401313, -0.18757272, -0.81094525,  0.75129018,  0.00972362,
        -0.62972637,  2.29536229,  0.0920418 ,  2.39252799,  0.54983815],
       [ 2.30868478,  1.30337355,  1.82463398,  2.67392052,  1.97367255,
         0.77750764,  1.81206697,  2.26971361,  3.17563985,  2.87988651],
       [ 3.03636353,  5.12093014,  5.79686488,  3.65529809,  2.50594319,
         2.84124027,  1.95026215,  7.47708735,  3.61420599,  1.65752765],
       [ 2.69287214,  3.46363202,  1.86159445,  1.66746094,  1.69755173,
         4.54410559,  1.51874636,  2.61081203,  5.10773362,  1.59635987]])

Формирование транспонированных матриц парных признаков (XT1, XT2, XT3, XT4, XT5, XT6, XT7, XT8, XT9, XT10, XT11, XT12):

In [ ]:
XT1 = np.array([XT[0], XT[1]])
XT2 = np.array([XT[1], XT[0]])
XT3 = np.array([XT[0], XT[2]])
XT4 = np.array([XT[2], XT[0]])
XT5 = np.array([XT[0], XT[3]])
XT6 = np.array([XT[3], XT[0]])
XT7 = np.array([XT[1], XT[2]])
XT8 = np.array([XT[2], XT[1]])
XT9 = np.array([XT[1], XT[3]])
XT10 = np.array([XT[3], XT[1]])
XT11 = np.array([XT[2], XT[3]])
XT12 = np.array([XT[3], XT[2]])

Транспонирование транспонированных матриц парных признаков в матрицы X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12 для дальнейшего решения задачи:

In [ ]:
X1 = np.transpose(XT1)
X2 = np.transpose(XT2)
X3 = np.transpose(XT3)
X4 = np.transpose(XT4)
X5 = np.transpose(XT5)
X6 = np.transpose(XT6)
X7 = np.transpose(XT7)
X8 = np.transpose(XT8)
X9 = np.transpose(XT9)
X10 = np.transpose(XT10)
X11 = np.transpose(XT11)
X12 = np.transpose(XT12)

Добавление единичного признака для приведения матриц праных признаков в удобную для дальнейшего решения задачи форму:

In [ ]:
ones = np.ones((X1.shape[0], 1))
X1 = np.hstack((ones, X1))
X2 = np.hstack((ones, X2))
X3 = np.hstack((ones, X3))
X4 = np.hstack((ones, X4))
X5 = np.hstack((ones, X5))
X6 = np.hstack((ones, X6))
X7 = np.hstack((ones, X7))
X8 = np.hstack((ones, X8))
X9 = np.hstack((ones, X9))
X10 = np.hstack((ones, X10))
X11 = np.hstack((ones, X11))
X12 = np.hstack((ones, X12))

Создание моделей линейной регресии для каждой пары признаков (lr1, lr2, lr3, lr4, lr5, lr6, lr7, lr8, lr9, lr10, lr11, lr12):

In [ ]:
lr1 = LinearRegression(fit_intercept=False).fit(X1, y)
lr2 = LinearRegression(fit_intercept=False).fit(X2, y)
lr3 = LinearRegression(fit_intercept=False).fit(X3, y)
lr4 = LinearRegression(fit_intercept=False).fit(X4, y)
lr5 = LinearRegression(fit_intercept=False).fit(X5, y)
lr6 = LinearRegression(fit_intercept=False).fit(X6, y)
lr7 = LinearRegression(fit_intercept=False).fit(X7, y)
lr8 = LinearRegression(fit_intercept=False).fit(X8, y)
lr9 = LinearRegression(fit_intercept=False).fit(X9, y)
lr10 = LinearRegression(fit_intercept=False).fit(X10, y)
lr11 = LinearRegression(fit_intercept=False).fit(X11, y)
lr12 = LinearRegression(fit_intercept=False).fit(X12, y)

Коэффициенты моделей линейной регрессии для каждой пары признаков (b1, b2, b3, b4, b5, b6, b7, b8, b9, b10, b11, b12):

In [ ]:
b1 = lr1.coef_
b2 = lr2.coef_
b3 = lr3.coef_
b4 = lr4.coef_
b5 = lr5.coef_
b6 = lr6.coef_
b7 = lr7.coef_
b8 = lr8.coef_
b9 = lr9.coef_
b10 = lr10.coef_
b11 = lr11.coef_
b12 = lr12.coef_

Массивы значений модели линейной регрессии для каждой пары признаков (Z1, Z2, Z3, Z4, Z5, Z6, Z7, Z8, Z9, Z10, Z11, Z12):

In [ ]:
Z1 = X1.dot(b1)
Z2 = X2.dot(b2)
Z3 = X3.dot(b3)
Z4 = X3.dot(b4)
Z5 = X3.dot(b5)
Z6 = X3.dot(b6)
Z7 = X3.dot(b7)
Z8 = X3.dot(b8)
Z9 = X3.dot(b9)
Z10 = X3.dot(b10)
Z11 = X3.dot(b11)
Z12 = X3.dot(b12)

Коэффициенты детерминации модели линейной регрессии для каждой пары признаков (r2_1, r2_2, r2_3, r2_4, r2_5, r2_6, r2_7, r2_8, r2_9, r2_10, r2_11, r2_12):

In [ ]:
r2_1 = np.corrcoef(y, Z1) ** 2
r2_2 = np.corrcoef(y, Z2) ** 2
r2_3 = np.corrcoef(y, Z3) ** 2
r2_4 = np.corrcoef(y, Z4) ** 2
r2_5 = np.corrcoef(y, Z5) ** 2
r2_6 = np.corrcoef(y, Z6) ** 2
r2_7 = np.corrcoef(y, Z7) ** 2
r2_8 = np.corrcoef(y, Z8) ** 2
r2_9 = np.corrcoef(y, Z9) ** 2
r2_10 = np.corrcoef(y, Z10) ** 2
r2_11 = np.corrcoef(y, Z11) ** 2
r2_12 = np.corrcoef(y, Z12) ** 2
r2 = r2_1[0][1], r2_2[0][1], r2_3[0][1], r2_4[0][1], r2_5[0][1], r2_6[0][1], r2_7[0][1], r2_8[0][1], r2_9[0][1], r2_10[0][1], r2_11[0][1], r2_12[0][1]
r2

(0.181135947425852,
 0.18113594742585226,
 0.7634246238793149,
 0.6625224959282909,
 0.6103431121601612,
 0.7445923886559904,
 0.7159421858650817,
 0.5811680634272685,
 0.7633241246936724,
 0.6664547500217303,
 0.6370360161828557,
 0.759076337903573)

Максимальный коэффициент детерминации модели линейной регрессии:

In [ ]:
max(r2)

0.7634246238793149

Ответ: 0.76342462.

### Задача 2

__Постановка задачи__

Является ли значимым уравнение регрессии, полученное в предыдущей задаче? В качестве ответа запишите 1, если является, и 0 иначе.

__Решение__

Максимальный коэффициент детерминации модели линейной регрессии из прошлой задачи (r2):

In [ ]:
r2 = r2_3[0][1]
r2

0.7634246238793149

Параметры статистики F-теста Фишера для матрицы парных признаков (X3) с максимальным коэффициентом детерминации модели линейной регрессии из прошлой задачи (n, m, k1, k2):

In [ ]:
n = X3.shape[0]
m = X3.shape[1] - 1  # поскольку добавляли нулевой фактор
k1 = m
k2 = n - m - 1
n, m, k1, k2

(10, 2, 2, 7)

Критическое значение статистки (t):

In [ ]:
alpha = 0.05
t = stats.f.ppf(1 - alpha, k1, k2)
t

4.73741412777588

Значение статистики (F):

In [ ]:
F = (r2 / k1) / ((1 - r2) / k2)
F

11.294439122922629

Поскольку значение стаитсики (F) превышает критическое значение статистики (t), то уравнение регресии принимается значимым. 

Ответ: 1.

### Задача 3

__Постановка задачи__

Для проведения A/B-тестирования сайта интернет-магазина были получены следующие данные: страница `A` была посещена 2509 раз, из них 77 закончились совершением покупки, страница `B` была посещена 1465 раз, 60 из них закончились совершением покупки. Является ли значимым отличие конверсии на страницах `A` и `B`? В качестве ответа запишите 1, если является, и 0 иначе.
_Подсказка_. Реализуйте двухвыборочный t-тест. В качестве выборок здесь можно взять наборы меток совершения покупки (0 или 1) каждым посетителем.

__Решение__

Пустые массивы выборок для страницы A (a) и страницы B (b):

In [ ]:
a = np.array([0 for i in range(2509)])
b = np.array([0 for i in range(1465)])

Заполнение массивов выборок a и b единичными значениями в соотвествии с количеством совершённых покупок:

In [ ]:
a[:77] = 1
b[:60] = 1

Размеры выборок для страниц A и B соотвественно (na, nb):

In [ ]:
na = a.size
nb = b.size
na, nb

(2509, 1465)

Несчещённые оценки стандарных отклонений выборок для страниц A и B соотвественно (sa, sb):

In [ ]:
sa = a.std(ddof=1)
sb = b.std(ddof=1)
sa, sb

(0.17250951379621338, 0.19825513126725994)

Cреднее квадратическое отклонение разности между выборочными средними выборок a и b:

In [ ]:
s_delta = np.sqrt(sa ** 2 / na + sb ** 2 / nb)
s_delta

0.006220171278295827

Значение статистики (t):

In [ ]:
t = (a.mean() - b.mean()) / s_delta
t

-1.6504551408398205

Число степеней свободы (df):

In [ ]:
df = (sa ** 2 / na + sb ** 2 / nb) ** 2 / \
    ((sa ** 2 / na) ** 2 / (na - 1) + (sb ** 2 / nb) ** 2 / (nb - 1))
df

2732.8025644352133

Квантили (t1, t2):

In [ ]:
alpha = 0.05
t1 = stats.t.ppf(alpha / 2, df=df)
t2 = stats.t.ppf(1 - alpha / 2, df=df)
t1, t2

(-1.9608324352746576, 1.9608324352746571)

Поскольку значение статистики не входит в критическую область, отсекаемую квантилями, то отличие конверсии на страницах A и B принимается не значимым.

Ответ: 0.